src/Compilers/Core/Desktop/CommandLine/ErrorLogger.cs

```diff
@@ -0,0 +1,372 @@
+
+            var toolInfo = GetToolInfo(toolName, toolFileVersion, toolAssemblyVersion);
+            WriteKeyValuePair(WellKnownStrings.ToolInfo, toolInfo, isFirst: false);
+
+            WriteKey(WellKnownStrings.Issues, isFirst: false);
+            StartList();
+        }
+
+        private Value GetToolInfo(string toolName, string toolFileVersion, Version toolAssemblyVersion)
+        {
+            var builder = ArrayBuilder<KeyValuePair<string, Value>>.GetInstance();
+            builder.Add(CreateSimpleKeyValuePair(WellKnownStrings.ToolName, toolName));
+            builder.Add(CreateSimpleKeyValuePair(WellKnownStrings.ToolAssemblyVersion, toolAssemblyVersion.ToString(fieldCount: 3)));
+            builder.Add(CreateSimpleKeyValuePair(WellKnownStrings.ToolFileVersion, GetToolFileVersionSubStr(toolFileVersion)));
+            return Value.Create(builder.ToImmutableAndFree(), this);
+        }
+        
+        private string GetToolFileVersionSubStr(string toolFileVersion)
+        {
+            // Our log format specifies that fileVersion can have at most 3 fields.
+
+            int count = 0;
+            for (int i = 0; i < toolFileVersion.Length; i++)
+            {
+                if (toolFileVersion[i] == '.')
+                {
+                    if (count == 2)
+                    {
+                        return toolFileVersion.Substring(0, i);
+                    }
+
+                    count++;
+                }
+            }
+
+            return toolFileVersion;
+        }
+
+        internal static void LogDiagnostic(Diagnostic diagnostic, CultureInfo culture, ErrorLogger errorLogger)
+        {
+            if (errorLogger != null)
+            {
+#pragma warning disable RS0013 // We need to invoke Diagnostic.Descriptor here to log all the metadata properties of the diagnostic.
+                var issue = new Issue(diagnostic.Id, diagnostic.GetMessage(culture),
+                    diagnostic.Descriptor.Description.ToString(culture), diagnostic.Descriptor.Title.ToString(culture),
+                    diagnostic.Category, diagnostic.Descriptor.HelpLinkUri, diagnostic.IsEnabledByDefault,
+                    diagnostic.DefaultSeverity, diagnostic.Severity, diagnostic.WarningLevel, diagnostic.Location,
+                    diagnostic.AdditionalLocations, diagnostic.CustomTags, diagnostic.Properties);
+#pragma warning restore RS0013
+
+                errorLogger.LogIssue(issue);
+            }
+        }
+
+        internal static void LogDiagnostic(DiagnosticInfo diagnostic, CultureInfo culture, ErrorLogger errorLogger)
+        {
+            if (errorLogger != null)
+            {
+#pragma warning disable RS0013 // We need to invoke Diagnostic.Descriptor here to log all the metadata properties of the diagnostic.
+                var issue = new Issue(diagnostic.MessageIdentifier, diagnostic.GetMessage(culture),
```

Comment:
I was primarily doing this because we wanted to log both Diagnostic and DiagnosticInfo, but as noted in the previous comment, this can be simplified by invoking Diagnostic.Create(diagnosticInfo) and avoid the Issue type completely. I'll make that change.